src/fastpath/fastpath.c

```diff
@@ -6,15 +6,313 @@
+#ifdef ENABLE_SMP_SUPPORT
+#ifdef CONFIG_HAVE_FPU
+                if (nativeThreadUsingFPU(dest)) {
+                    slowpath(SysSend);
+                    UNREACHABLE();
+                }
+#endif
+#endif
+            } else {
+                sc = dest->tcbSchedContext;
+            }
+
+            /* Signal to higher prio thread is NOT fastpathed. Not sure if this handles idle thread */
+            if (NODE_STATE_ON_CORE(ksCurThread, sc->scCore)->tcbPriority < dest->tcbPriority) {
+                slowpath(SysSend);
+            }
+
+            /* Simplified schedContext_resume that does not change state and reverts to the
+            * slowpath in cases where the SC does not have sufficient budget, as this case
+            * adds extra scheduler logic. Normally, this is done after the sched_context donate
+            * but after tweaking it, I don't immediately see anything executed in schedContext_donate
+            * that will affect the conditions of this check */
+
+            if (sc->scRefillMax > 0 && !thread_state_get_tcbInReleaseQueue(dest->tcbState)) {
+                if (!(refill_ready(sc) && refill_sufficient(sc, 0))) {
+                    slowpath(SysSend);
+                    UNREACHABLE();
+                }
+                schedulable = true;
+            }
+
+            if (ksCurDomain != dest->tcbDomain SMP_COND_STATEMENT( || sc->scCore != getCurrentCPUIndex())) {
+                crossnode = true;
+            }
+
+            /*  Point of no return */
+#ifdef CONFIG_BENCHMARK_TRACK_KERNEL_ENTRIES
+            ksKernelEntry.is_fastpath = true;
+#endif
+
+            /* Equivalent to cancel_ipc */
+            endpoint_t *ep_ptr;
+            tcb_queue_t queue;
+            ep_ptr = EP_PTR(thread_state_get_blockingObject(dest->tcbState));
+
+            queue = ep_ptr_get_queue(ep_ptr);
+            queue = tcbEPDequeue(dest, queue);
+            ep_ptr_set_queue(ep_ptr, queue);
+
+            if (!queue.head) {
+                endpoint_ptr_set_state(ep_ptr, EPState_Idle);
+            }
+
+            reply_t *reply = REPLY_PTR(thread_state_get_replyObject(dest->tcbState));
+            if (reply != NULL) {
+                reply_unlink(reply, dest);
+            }
+
+            /* Now we donate the SC. The checks required for this were
+             * already done before the point of no return. */
+            setRegister(dest, badgeRegister, badge);
+            thread_state_ptr_set_tsType_np(&dest->tcbState, ThreadState_Running);
+
+            if (!dest->tcbSchedContext) {
+                /* Equivalent to schedContext_donate without migrateTCB() */
+                sc->scTcb = dest;
+                dest->tcbSchedContext = sc;
+            }
+            assert(dest->tcbSchedContext);
+
+#ifdef ENABLE_SMP_SUPPORT
+            /* The part of migrateTCB() that doesn't involve the slowpathed FPU save */
+            dest->tcbAffinity = sc->scCore;
```

Comment:
This will also need the updates to the debug TCB queues on either side of the affinity update:

```
#ifdef CONFIG_DEBUG_BUILD
    tcbDebugRemove(tcb);
#endif
    dest->tcbAffinity = sc->scCore;
#ifdef CONFIG_DEBUG_BUILD
    tcbDebugAppend(tcb);
#endif
```